source/dofs/dof_handler.cc

```diff
@@ -2643,6 +2643,99 @@ DoFHandler<dim, spacedim>::distribute_mg_dofs()
+
+  if (!hp_capability_enabled)
+    // nothing to do
+    return false;
+
+  // communication of future fe indices on ghost cells necessary.
+  // thus, currently only implemented for serial triangulations.
+  Assert((dynamic_cast<const parallel::TriangulationBase<dim, spacedim> *>(
+            &(*tria)) == nullptr),
+         ExcNotImplemented());
+
+  //
+  // establish hierarchy
+  //
+  // - create bimap between hierarchy levels and FE indices
+  // - classify cells based on their level
+  const std::vector<unsigned int> fe_on_level =
+    fe_collection.get_hierarchy_sequence(contains_fe_index);
+
+  std::map<unsigned int, unsigned int> level_on_fe;
+  for (unsigned int i = 0; i < fe_on_level.size(); ++i)
+    level_on_fe[fe_on_level[i]] = i;
+
+  std::vector<std::list<active_cell_iterator>> cells_on_level(
+    fe_on_level.size());
+  if (fe_on_level.size() == fe_collection.size())
+    {
+      // all FE indices are part of hierarchy
+      for (const auto &cell : active_cell_iterators())
+        cells_on_level[level_on_fe[cell->future_fe_index()]].push_back(cell);
+    }
+  else
+    {
+      // some FE indices are part of hierarchy
+      typename decltype(level_on_fe)::iterator iter;
+      for (const auto &cell : active_cell_iterators())
+        if ((iter = level_on_fe.find(cell->future_fe_index())) !=
+            level_on_fe.end())
+          cells_on_level[iter->second].push_back(cell);
+    }
+
+  //
+  // limit level difference of neighboring cells
+  //
+  // - always raise levels to match criterion, never lower them
+  //   (hence iterating from highest to lowest level)
+  // - update future FE indices
+  // - move cells in level representation correspondigly
+  bool changed_fe_indices = false;
+  for (unsigned int level = fe_on_level.size() - 1; level > max_difference;
+       --level)
+    for (const auto &cell : cells_on_level[level])
+      for (unsigned int f = 0; f < cell->n_faces(); ++f)
+        if (cell->face(f)->at_boundary() == false)
+          {
+            const auto neighbor = cell->neighbor(f);
+
+            const unsigned int neighbor_level =
+              level_on_fe[neighbor->future_fe_index()];
+
+            if ((level - max_difference) > neighbor_level)
+              {
+                // remove neighbor from old level representation
+                auto &     old_cells = cells_on_level[neighbor_level];
+                const auto iter =
+                  std::find(old_cells.begin(), old_cells.end(), neighbor);
+                old_cells.erase(iter);
+
+                // move neighbor to new level representation
+                const unsigned int new_level = level - max_difference;
+                cells_on_level[new_level].push_back(neighbor);
+
+                // update future FE index
+                neighbor->set_future_fe_index(fe_on_level[new_level]);
+                changed_fe_indices = true;
+              }
+          }
+
+  return changed_fe_indices;
```

Comment:
I am not sure if it wouldn't be simpler to work on a single `vector` containing the level and having the size of active cells. Actually I think, we could introduce a global vector and use `Triangulation::global_active_cell_index_partitioner()`/`CellAccessor::global_active_cell_index()` (see also https://github.com/dealii/dealii/blob/5c1822a561be1238e0a51f155a6dcf78ca531904/tests/grid/global_ids_01.cc#L43-L48). Working on a vector and updating it is simple. @kronbichler and I wrote these function for such purposes (`MatrixFree` in mind) but we didn't have time to use it yet ourselves (except in one place indirectly: https://github.com/dealii/dealii/blob/b00679e4089197fd1c0c5e7a2210105d21cff9ea/include/deal.II/matrix_free/matrix_free.templates.h#L1748-L1751).

Reply:
I don't feel strongly about what the best approach is. But the update of ghost cells is easy and efficient in this case since the communication pattern has already be setup once (also using a single `GridTools::exchange_cell_data_to_ghosts`, which is called anyway here: https://github.com/dealii/dealii/blob/e694e602704f0c418cd81484972628ca26f33ab2/source/distributed/tria_base.cc#L411-L417).

Reply:
I'll give the global vector a try. But let's tackle this in a follow-up PR!